packages/xstate-immer/src/index.ts

```diff
@@ -61,3 +63,52 @@ export function createUpdater<TContext, TEvent extends ImmerUpdateEvent>(
     type
   };
 }
+
+const IMMER_PATCH_TYPE = '@immer.patch' as const;
+
+export interface ImmerPatchEvent<K> {
+  type: typeof IMMER_PATCH_TYPE;
+  key: K;
+  patches: Patch[];
+}
+
+export const createPatch = <T, K extends keyof T>(
+  key: K
+): {
+  (currentValue: T[K], fn: (draft: Draft<T[K]>) => void): ImmerPatchEvent<K>;
+  type: ImmerPatchEvent<K>['type'];
+  action: AssignAction<T, ImmerPatchEvent<K>>;
+  transition: Record<
+    ImmerPatchEvent<K>['type'],
+    { actions: Array<AssignAction<T, ImmerPatchEvent<K>>> }
+  >;
+} => {
+  const applyPatch = (
+    currentValue: T[K],
+    fn: (draft: Draft<T[K]>) => void
+  ): ImmerPatchEvent<K> => {
+    const patches: Patch[] = [];
+
+    produce(currentValue, fn, (resultPatches) => {
+      patches.push(...resultPatches);
+    });
+
+    return {
+      type: IMMER_PATCH_TYPE,
+      key,
+      patches
+    };
+  };
+
+  const action = xstateAssign({
+    [key as K]: (ctx: T, e: ImmerPatchEvent<K>) => {
+      return applyPatches(ctx[key], e.patches);
+    }
+  }) as AssignAction<T, ImmerPatchEvent<K>>;
+
+  applyPatch.type = IMMER_PATCH_TYPE;
+  applyPatch.action = action;
+  applyPatch.transition = { [IMMER_PATCH_TYPE]: { actions: [action] } };
```

Comment:
This is IMHO problematic for a couple of reasons:
1. returning `type` from here makes it look like this is a type exclusive for this "patcher", but it is not - it's a "global" type
2. and because it's a global type the composition suffers right now because it's not obvious (if not impossible) how to create 2 "patchers" and apply them at the same state node.

The whole thing is very similar to `createUpdater` and I believe that the API should be shared as much as possible between both of them (I'm not super familiar with Immer, so not sure how much it is possible to share the API, but at least they should resemble each other).

A global type could be used (but not returned from this factory) if only we could produce patches based on the "root" context - then all could just be applied uniformly based on a single event type. This is just a random thought - not sure how feasible this even is.

Reply:
That's a fair assumption, we can probably construct different types per "patcher"